src/index.mjs

```diff
@@ -0,0 +1,314 @@
+* 
+* @type {Object}
+* @property {number} 0 - Zero-indexed position of the "type" value in the internal event format
+* @property {number} 1 - Zero-indexed position of the "name" value in the internal event format
+* @property {number} 2 - Zero-indexed position of the "createdAt" value in the internal event format
+* @property {number} 3 - Zero-indexed position of the "size" value in the internal event format
+* @property {number} 4 - Zero-indexed position of the "nblocks" value in the internal event format
+@default
+*/
+const EVENT_MAPPING = {
+  0: "type",
+  1: "name",
+  2: "createdAt",
+  3: "size",
+  4: "nblocks",
+};
+
+/** @constant
+* Mapping of fields that are integers and need parsing when deserialized
+* @type {Object}
+* @property {string} "createdAt" - A unix timestamp recording when the event was recorded
+* @property {string} "size" - Size in bytes of memory allocated by the njs vm
+* @property {string} "nblocks" - Number of memory blocks allocated by the njs vm
+@default
+*/
+const INT_FIELDS = {
+  createdAt: true,
+  size: true,
+  nblocks: true,
+};
+
+/** @constant
+* Enums for the status of the profiler.  Used to prevent multiple initializations/collections
+* Enums as strings to avoid unnecessary parsing when taking out of nginx variables
+* @type {Object}
+* @property {string} "RUNNING" - Enum indicating that the profiler has been initialized
+* @property {string} "DONE" - Enum indicating that the profiler is done for the request
+@default
+*/
+const PROFILER_STATUS = {
+  RUNNING: "1",
+  DONE: "2",
+};
+
+/**
+ * Starts the profiler. You can provide an arbitrary id to indentify the profile.
+ * If not given, one will be generated for you.
+ *
+ * @param {NJSRequest} r - The njs request object for variable access
+ * @param {string} requestId - Arbitrary id to indentify the profile. Defaults to a value generated by `request_id`
+ * @param {function} reporterFn - One of the reporter functions exported by this module. Defaults to `logReporter`
+ *
+ * @example
+ *
+ *    import profiler from "./njs_modules/njs-memory-profiler/njs-memory-profiler.js";
+ *    profiler.init(r, "my_id", profiler.fileReporter);
+ */
+function init(r, requestId, reporterFn) {
+  // Don't init more than once
+  if (r.variables.profiler_status === PROFILER_STATUS.RUNNING) return;
+
+  // You can specify your own id for the request profile, or the profiler will generate one
+  r.variables.profiler_request_id = requestId || r.variables.request_id;
+  reporterFn = reporterFn || logReporter;
+
+  pushEvent(r, EVENT_TYPE_TO_STRING[`${EVENT_TYPES.START}`], EVENT_TYPES.START);
+  r.variables.profiler_status = PROFILER_STATUS.RUNNING;
+
+  // We run final collection on vm exit, but you can manually
+  // invoke collection earlier by manually calling `collect`.
+  // If this is done, we don't collect again on vm exit.
+  njs.on("exit", () => {
+    // No async work in this context
+    collect(r, reporterFn);
```

Comment:
In the docs it says `collect` must be called manually, does this remove that requirement?

Reply:
I see you got it below, but let me clarify that in the readme since if it confused you it will confuse others.